hudi-common/src/main/java/org/apache/hudi/common/model/HoodiePartitionMetadata.java

```diff
@@ -117,30 +133,119 @@ public void trySave(int taskPartitionId) {
+          // parameters are not important.
+          OrcFile.WriterOptions writerOptions = OrcFile.writerOptions(fs.getConf()).fileSystem(fs)
+              .setSchema(AvroOrcUtils.createOrcSchema(schema));
+          try (Writer writer = OrcFile.createWriter(filePath, writerOptions)) {
+            for (String key : props.stringPropertyNames()) {
+              writer.addUserMetadata(key, ByteBuffer.wrap(props.getProperty(key).getBytes()));
+            }
+          }
+          break;
+        default:
+          throw new HoodieException("Unsupported format for partition metafiles: " + format.get());
+      }
+    } else {
+      // Backwards compatible properties file format
+      FSDataOutputStream os = fs.create(filePath, true);
+      props.store(os, "partition metadata");
+      os.hsync();
+      os.hflush();
+      os.close();
+    }
+  }
+
   /**
    * Read out the metadata for this partition.
    */
   public void readFromFS() throws IOException {
-    FSDataInputStream is = null;
+    Option<Path> metafilePath = getPartitionMetafilePath(fs, partitionPath);
+    if (!metafilePath.isPresent()) {
+      throw new HoodieException("Partition metafile not found in path " + partitionPath);
+    }
+
     try {
-      Path metaFile = new Path(partitionPath, HOODIE_PARTITION_METAFILE);
-      is = fs.open(metaFile);
-      props.load(is);
-    } catch (IOException ioe) {
-      throw new HoodieException("Error reading Hoodie partition metadata for " + partitionPath, ioe);
-    } finally {
-      if (is != null) {
-        is.close();
+      BaseFileUtils reader = BaseFileUtils.getInstance(metafilePath.toString());
+      Map<String, String> metadata = reader.readFooter(fs.getConf(), true, metafilePath.get(), PARTITION_DEPTH_KEY, COMMIT_TIME_KEY);
+      props.clear();
+      metadata.forEach((k, v) -> props.put(k, v));
+    } catch (UnsupportedOperationException e) {
+      // Properties file format
+      FSDataInputStream is = null;
+      try {
+        is = fs.open(metafilePath.get());
+        props.load(is);
+      } catch (IOException ioe) {
+        throw new HoodieException("Error reading Hoodie partition metadata from " + metafilePath, ioe);
+      } finally {
+        if (is != null) {
+          is.close();
+        }
       }
     }
   }
 
   // methods related to partition meta data
   public static boolean hasPartitionMetadata(FileSystem fs, Path partitionPath) {
+    return getPartitionMetafilePath(fs, partitionPath).isPresent();
+  }
+
+  /**
+   * Returns the name of the partition metadata.
+   *
+   * @param fs
+   * @param partitionPath
+   * @return Name of the partition metafile or empty option
+   */
+  public static Option<Path> getPartitionMetafilePath(FileSystem fs, Path partitionPath) {
+    // The partition listing is a costly operation so instead we are searching for existence of the files instead.
+    // This is in expected order as properties file based partition metafiles should be the most common.
```

Comment:
can't we read the table config and determine what to do? e.g HoodieTableConfig.PARTITION_METAFILE_USE_DATA_FORMAT=true, then check for base file format, else read a txt properties file

Reply:
There is a catch22 here: We are trying to get the path to partition meta file so we can read it to get the partition depth to reach the .hoodie folder.
So within this function, we cannot get the tableConfig which requires us to know the basePath to the .hoodie folder.

Reply:
duh. you are right. I still like to avoid the exists call. let me think about it little more.

Reply:
If we can upgrade all the files as a part of upgrade/downgrade and the readers/query engines are pre upgraded, then this could work. otherwise, we need to do something like what you are doing.